/* PROGRAM PATH - DO NOT CHANGE MANUALLY THIS FILE -- It is generated automatically by the build process -- See the 'update' target in the 'makefile' for more details */

//ERROR MESSAGES


#define ADDRESS_BUILT_IN "built_in"
#define PROGRAM_BUILT_IN "t_coffee"
#define TEST_WWWSITE_4_TCOFFEE "www.google.com"


//TclinkdbStart

#define TCOFFEE_4_TCOFFEE "t_coffee"
#define TCOFFEE_type "sequence_multiple_aligner"
#define TCOFFEE_ADDRESS "http://www.tcoffee.org"
#define TCOFFEE_language "C++"
#define TCOFFEE_language2 "CXX"
#define TCOFFEE_source "http://www.tcoffee.org/Packages/sources/tcoffee/stable/T-COFFEE_distribution.tar.gz"
#define TCOFFEE_update_action "always"
#define TCOFFEE_binary "t_coffee"
#define TCOFFEE_mode "tcoffee,mcoffee,rcoffee,expresso,3dcoffee"
#define CLUSTALO_4_TCOFFEE "clustalo"
#define CLUSTALO_type "sequence_multiple_aligner"
#define CLUSTALO_ADDRESS "http://www.clustal.org/omega/"
#define CLUSTALO_language "C++"
#define CLUSTALO_language2 "C++"
#define CLUSTALO_source "http://www.clustal.org/omega/clustal-omega-1.2.4.tar.gz"
#define CLUSTALO_mode "mcoffee"
#define CLUSTALO_binary "clustalo"
#define CLUSTALO_version "1.2.4"
#define STRIKE_4_TCOFFEE "strike"
#define STRIKE_type "sequence_alignment_scoring"
#define STRIKE_ADDRESS "http://www.tcoffee.org/Projects/strike/index.html"
#define STRIKE_language "C++"
#define STRIKE_language2 "CXX"
#define STRIKE_source "http://www.tcoffee.org/Projects/strike/strike_v1.2.tar.bz2"
#define STRIKE_mode "tcoffee,expresso"
#define STRIKE_version "1.2"
#define STRIKE_binary "strike"
#define CLUSTALW2_4_TCOFFEE "clustalw2"
#define CLUSTALW2_type "sequence_multiple_aligner"
#define CLUSTALW2_ADDRESS "http://www.clustal.org"
#define CLUSTALW2_language "C++"
#define CLUSTALW2_language2 "CXX"
#define CLUSTALW2_source "http://www.clustal.org/download/2.0.10/clustalw-2.0.10-src.tar.gz"
#define CLUSTALW2_mode "mcoffee,rcoffee"
#define CLUSTALW2_binary "clustalw2"
#define CLUSTALW2_version "2.0.10"
#define CLUSTALW_4_TCOFFEE "clustalw"
#define CLUSTALW_type "sequence_multiple_aligner"
#define CLUSTALW_ADDRESS "http://www.clustal.org"
#define CLUSTALW_language "C"
#define CLUSTALW_language2 "C"
#define CLUSTALW_source "http://www.clustal.org/download/1.X/ftp-igbmc.u-strasbg.fr/pub/ClustalW/clustalw1.82.UNIX.tar.gz"
#define CLUSTALW_mode "mcoffee,rcoffee"
#define CLUSTALW_version "1.82"
#define CLUSTALW_binary "clustalw"
#define DIALIGNT_4_TCOFFEE "dialign-t"
#define DIALIGNT_type "sequence_multiple_aligner"
#define DIALIGNT_ADDRESS "http://dialign-tx.gobics.de/"
#define DIALIGNT_DIR "/usr/share/dialign-tx/"
#define DIALIGNT_language "C"
#define DIALIGNT_language2 "C"
#define DIALIGNT_source "http://dialign-tx.gobics.de/DIALIGN-TX_1.0.2.tar.gz"
#define DIALIGNT_mode "mcoffee"
#define DIALIGNT_binary "dialign-t"
#define DIALIGNT_version "1.0.2"
#define DIALIGNTX_4_TCOFFEE "dialign-tx"
#define DIALIGNTX_type "sequence_multiple_aligner"
#define DIALIGNTX_ADDRESS "http://dialign-tx.gobics.de/"
#define DIALIGNTX_DIR "/usr/share/dialign-tx/"
#define DIALIGNTX_language "C"
#define DIALIGNTX_language2 "C"
#define DIALIGNTX_source "http://dialign-tx.gobics.de/DIALIGN-TX_1.0.2.tar.gz"
#define DIALIGNTX_mode "mcoffee"
#define DIALIGNTX_binary "dialign-tx"
#define DIALIGNTX_version "1.0.2"
#define POA_4_TCOFFEE "poa"
#define POA_type "sequence_multiple_aligner"
#define POA_ADDRESS "http://www.bioinformatics.ucla.edu/poa/"
#define POA_language "C"
#define POA_language2 "C"
#define POA_source "http://downloads.sourceforge.net/poamsa/poaV2.tar.gz"
#define POA_DIR "/usr/share/"
#define POA_FILE1 "blosum80.mat"
#define POA_mode "mcoffee"
#define POA_binary "poa"
#define POA_version "2.0"
#define PROBCONS_4_TCOFFEE "probcons"
#define PROBCONS_type "sequence_multiple_aligner"
#define PROBCONS_ADDRESS "http://probcons.stanford.edu/"
#define PROBCONS_language2 "CXX"
#define PROBCONS_language "C++"
#define PROBCONS_source "http://probcons.stanford.edu/probcons_v1_12.tar.gz"
#define PROBCONS_mode "mcoffee"
#define PROBCONS_binary "probcons"
#define PROBCONS_version "1.12"
#define MSAPROBS_4_TCOFFEE "msaprobs"
#define MSAPROBS_type "sequence_multiple_aligner"
#define MSAPROBS_ADDRESS "http://msaprobs.sourceforge.net/homepage.htm#latest"
#define MSAPROBS_language2 "CXX"
#define MSAPROBS_language "C++"
#define MSAPROBS_source "https://sourceforge.net/projects/msaprobs/files/MSAProbs-MPI/MSAProbs-MPI_rel1.0.5.tar.gz"
#define MSAPROBS_mode "mcoffee"
#define MSAPROBS_binary "msaprobs"
#define MSAPROBS_version "1.05"
#define MSAPROBS_update_action "never"
#define UPP_4_TCOFFEE "upp"
#define UPP_type "sequence_multiple_aligner"
#define UPP_ADDRESS "http://www.cs.utexas.edu/users/phylo/software/upp/"
#define UPP_language2 "CXX"
#define UPP_language "C++"
#define UPP_source "https://github.com/smirarab/pasta/archive/upp.zip"
#define UPP_mode "mcoffee"
#define UPP_binary "upp"
#define UPP_version "1"
#define UPP_update_action "never"
#define FAMSA_4_TCOFFEE "famsa"
#define FAMSA_type "sequence_multiple_aligner"
#define FAMSA_ADDRESS "https://github.com/refresh-bio/FAMSA"
#define FAMSA_language "C++"
#define FAMSA_language "C++"
#define FAMSA_source "https://github.com/refresh-bio/FAMSA.git"
#define FAMSA_mode "mcoffee,rcoffee"
#define FAMSA_binary "famsa"
#define FAMSA_version "1.1"
#define MAFFT_4_TCOFFEE "mafft"
#define MAFFT_type "sequence_multiple_aligner"
#define MAFFT_ADDRESS "http://align.bmr.kyushu-u.ac.jp/mafft/online/server/"
#define MAFFT_language "C"
#define MAFFT_language "C"
#define MAFFT_source "http://mafft.cbrc.jp/alignment/software/mafft-7.310-with-extensions-src.tgz"
#define MAFFT_mode "mcoffee,rcoffee"
#define MAFFT_binary "mafft.tar.gz"
#define MAFFT_version "7.310"
#define MSA_4_TCOFFEE "msa"
#define MSA_type "sequence_multiple_aligner"
#define MSA_ADDRESS "https://www.ncbi.nlm.nih.gov/CBBresearch/Schaffer/msa.html"
#define MSA_language "C"
#define MSA_language "C"
#define MSA_source "ftp://ftp.ncbi.nih.gov/pub/msa/msa.tar.Z"
#define MSA_mode "mcoffee"
#define MSA_binary "msa.pl"
#define MSA_version "1.0"
#define MSA_update_action "never"
#define DCA_4_TCOFFEE "dca"
#define DCA_type "sequence_multiple_aligner"
#define DCA_ADDRESS "https://bibiserv2.cebitec.uni-bielefeld.de/dca"
#define DCA_language "C"
#define DCA_language "C"
#define DCA_source "https://bibiserv2.cebitec.uni-bielefeld.de/applications/dca/resources/downloads/dca-1.1-src.tar.gz"
#define DCA_mode "mcoffee"
#define DCA_binary "dca.pl"
#define DCA_version "1.1"
#define DCA_update_action "never"
#define MUSCLE_4_TCOFFEE "muscle"
#define MUSCLE_type "sequence_multiple_aligner"
#define MUSCLE_ADDRESS "http://www.drive5.com/muscle/"
#define MUSCLE_language "C++"
#define MUSCLE_language2 "GPP"
#define MUSCLE_source "http://www.drive5.com/muscle/downloads3.7/muscle3.7_src.tar.gz"
#define MUSCLE_windows "http://www.drive5.com/muscle/downloads3.7/muscle3.7_win32.zip"
#define MUSCLE_linux "http://www.drive5.com/muscle/downloads3.7/muscle3.7_linux_ia32.tar.gz"
#define MUSCLE_mode "mcoffee,rcoffee"
#define MUSCLE_version "3.7"
#define PCMA_4_TCOFFEE "pcma"
#define PCMA_type "sequence_multiple_aligner"
#define PCMA_ADDRESS "http://prodata.swmed.edu/pcma/pcma.php"
#define PCMA_language "C"
#define PCMA_language2 "C"
#define PCMA_source "http://prodata.swmed.edu/download/pub/PCMA/pcma.tar.gz"
#define PCMA_mode "mcoffee"
#define PCMA_version "1.0"
#define KALIGN_4_TCOFFEE "kalign"
#define KALIGN_type "sequence_multiple_aligner"
#define KALIGN_ADDRESS "http://msa.cgb.ki.se"
#define KALIGN_language "C"
#define KALIGN_language2 "C"
#define KALIGN_source "http://msa.cgb.ki.se/downloads/kalign/current.tar.gz"
#define KALIGN_mode "mcoffee"
#define KALIGN_version "1.0"
#define AMAP_4_TCOFFEE "amap"
#define AMAP_type "sequence_multiple_aligner"
#define AMAP_ADDRESS "http://bio.math.berkeley.edu/amap/"
#define AMAP_language "C++"
#define AMAP_language2 "CXX"
#define AMAP_source "https://github.com/mes5k/amap-align/archive/amap.zip"
#define AMAP_mode "mcoffee"
#define AMAP_version "2.0"
#define AMAP_update_action "never"
#define PRODA_4_TCOFFEE "proda"
#define PRODA_type "sequence_multiple_aligner"
#define PRODA_ADDRESS "http://proda.stanford.edu"
#define PRODA_language "C++"
#define PRODA_language2 "CXX"
#define PRODA_source "http://proda.stanford.edu/proda_1_0.tar.gz"
#define PRODA_mode "mcoffee"
#define PRODA_version "1.0"
#define PRANK_4_TCOFFEE "prank"
#define PRANK_type "sequence_multiple_aligner"
#define PRANK_ADDRESS "http://www.ebi.ac.uk/goldman-srv/prank/"
#define PRANK_language "C++"
#define PRANK_language2 "CXX"
#define PRANK_source "http://www.ebi.ac.uk/goldman-srv/prank/src/prank/prank.src.100802.tgz"
#define PRANK_mode "mcoffee"
#define PRANK_version "100303"
#define SAP_4_TCOFFEE "sap"
#define SAP_type "structure_pairwise_aligner"
#define SAP_ADDRESS "https://mathbio.crick.ac.uk/wiki/Software#SAP"
#define SAP_language "C"
#define SAP_language2 "C"
#define SAP_source "https://github.com/jkleinj/SAP/archive/v.1.1.3.tar.gz"
#define SAP_mode "expresso,3dcoffee"
#define SAP_version "1.1.3"
#define SAP_binary "sap"
#define TMALIGN_4_TCOFFEE "TMalign"
#define TMALIGN_type "structure_pairwise_aligner"
#define TMALIGN_ADDRESS "http://zhanglab.ccmb.med.umich.edu/TM-align/TMalign.f"
#define TMALIGN_language "Fortran"
#define TMALIGN_language2 "Fortran"
#define TMALIGN_source "http://zhanglab.ccmb.med.umich.edu/TM-align/TMalign.f"
#define TMALIGN_linux "http://zhanglab.ccmb.med.umich.edu/TM-align/TMalign_32.gz"
#define TMALIGN_mode "expresso,3dcoffee"
#define TMALIGN_version "2013.05.11"
#define MUSTANG_4_TCOFFEE "mustang"
#define MUSTANG_type "structure_pairwise_aligner"
#define MUSTANG_ADDRESS "http://lcb.infotech.monash.edu.au/mustang/"
#define MUSTANG_language "C++"
#define MUSTANG_language2 "CXX"
#define MUSTANG_source "http://lcb.infotech.monash.edu.au/mustang/mustang_v3.2.3.tgz"
#define MUSTANG_mode "expresso,3dcoffee"
#define MUSTANG_version "3.2.3"
#define LSQMAN_4_TCOFFEE "lsqman"
#define LSQMAN_type "structure_pairwise_aligner"
#define LSQMAN_ADDRESS "empty"
#define LSQMAN_language "empty"
#define LSQMAN_language2 "empty"
#define LSQMAN_source "empty"
#define LSQMAN_update_action "never"
#define LSQMAN_mode "expresso,3dcoffee"
#define ALIGN_PDB_4_TCOFFEE "align_pdb"
#define ALIGN_PDB_type "structure_pairwise_aligner"
#define ALIGN_PDB_ADDRESS "empty"
#define ALIGN_PDB_language "empty"
#define ALIGN_PDB_language2 "empty"
#define ALIGN_PDB_source "empty"
#define ALIGN_PDB_update_action "never"
#define ALIGN_PDB_mode "expresso,3dcoffee"
#define FUGUE_4_TCOFFEE "fugueali"
#define FUGUE_type "structure_pairwise_aligner"
#define FUGUE_ADDRESS "http://mizuguchilab.org/fugue/"
#define FUGUE_language "empty"
#define FUGUE_language2 "empty"
#define FUGUE_source "empty"
#define FUGUE_update_action "never"
#define FUGUE_mode "expresso,3dcoffee"
#define DALILITEc_4_TCOFFEE "dalilite.pl"
#define DALILITEc_type "structure_pairwise_aligner"
#define DALILITEc_ADDRESS "built_in"
#define DALILITEc_ADDRESS2 "http://www.ebi.ac.uk/Tools/webservices/services/dalilite"
#define DALILITEc_language "Perl"
#define DALILITEc_language2 "Perl"
#define DALILITEc_source "empty"
#define DALILITEc_update_action "never"
#define DALILITEc_mode "expresso,3dcoffee"
#define PROBCONSRNA_4_TCOFFEE "probconsRNA"
#define PROBCONSRNA_type "RNA_multiple_aligner"
#define PROBCONSRNA_ADDRESS "http://probcons.stanford.edu/"
#define PROBCONSRNA_language "C++"
#define PROBCONSRNA_language2 "CXX"
#define PROBCONSRNA_source "http://probcons.stanford.edu/probconsRNA.tar.gz"
#define PROBCONSRNA_mode "mcoffee,rcoffee"
#define PROBCONSRNA_version "1.0"
#define CONSAN_4_TCOFFEE "sfold"
#define CONSAN_type "RNA_pairwise_aligner"
#define CONSAN_ADDRESS "http://selab.janelia.org/software/consan/"
#define CONSAN_language "empty"
#define CONSAN_language2 "empty"
#define CONSAN_source "empty"
#define CONSAN_update_action "never"
#define CONSAN_mode "rcoffee"
#define RNAPLFOLD_4_TCOFFEE "RNAplfold"
#define RNAPLFOLD_type "RNA_secondarystructure_predictor"
#define RNAPLFOLD_ADDRESS "http://www.tbi.univie.ac.at/RNA/"
#define RNAPLFOLD_language "C"
#define RNAPLFOLD_language2 "C"
#define RNAPLFOLD_source "http://www.tbi.univie.ac.at/RNA/packages/source/ViennaRNA-2.1.9.tar.gz"
#define RNAPLFOLD_mode "rcoffee,"
#define RNAPLFOLD_binary "RNAplfold.tar.gz"
#define RNAPLFOLD_version "2.1.9"
#define PHYLIP_4_TCOFFEE "retree"
#define PHYLIP_type "Phylogeny"
#define PHYLIP_ADDRESS "http://evolution.gs.washington.edu/phylip/"
#define PHYLIP_language "C"
#define PHYLIP_language2 "C"
#define PHYLIP_source "http://www.tcoffee.org/Packages/mirrors/source/phylip-3.66.tar.gz"
#define PHYLIP_mode "trmsd,"
#define PHYLIP_binary "retree.tar.gz"
#define PHYLIP_version "3.66"
#define HMMTOP_4_TCOFFEE "hmmtop"
#define HMMTOP_type "protein_secondarystructure_predictor"
#define HMMTOP_ADDRESS "www.enzim.hu/hmmtop/"
#define HMMTOP_language "C"
#define HMMTOP_language2 "C"
#define HMMTOP_source "http://www.tcoffee.org/Packages/mirrors/hmmtop2.1.tgz"
#define HMMTOP_binary "hmmtop"
#define HMMTOP_update_action "never"
#define HMMTOP_mode "psicoffee"
#define HMMTOP_version "2.1"
#define GOR4_4_TCOFFEE "gorIV"
#define GOR4_type "protein_secondarystructure_predictor"
#define GOR4_ADDRESS "http://mig.jouy.inra.fr/logiciels/gorIV/"
#define GOR4_language "C"
#define GOR4_language2 "C"
#define GOR4_source "http://www.tcoffee.org/Packages/mirrors/GOR_IV.tar.gz"
#define GOR4_update_action "never"
#define GOR4_mode "tcoffee"
#define EBIWUBLASTc_4_TCOFFEE "wublast.pl"
#define EBIWUBLASTc_type "protein_homology_predictor"
#define EBIWUBLASTc_ADDRESS "built_in"
#define EBIWUBLASTc_ADDRESS2 "http://www.ebi.ac.uk/Tools/webservices/services/wublast"
#define EBIWUBLASTc_language "Perl"
#define EBIWUBLASTc_language2 "Perl"
#define EBIWUBLASTc_source "empty"
#define EBIWUBLASTc_update_action "never"
#define EBIWUBLASTc_mode "psicoffee,expresso,accurate"
#define EBIBLASTPGPc_4_TCOFFEE "blastpgp.pl"
#define EBIBLASTPGPc_type "protein_homology_predictor"
#define EBIBLASTPGPc_ADDRESS "built_in"
#define EBIBLASTPGPc_ADDRESS2 "http://www.ebi.ac.uk/Tools/webservices/services/blastpgp"
#define EBIBLASTPGPc_language "Perl"
#define EBIBLASTPGPc_language2 "Perl"
#define EBIBLASTPGPc_source "empty"
#define EBIBLASTPGPc_update_action "never"
#define EBIBLASTPGPc_mode "psicoffee,expresso,accurate"
#define blastall_4_TCOFFEE "blastall"
#define blastall_type "protein_homology_predictor"
#define blastall_ADDRESS "ftp://ftp.ncbi.nih.gov/blast/executables/LATEST"
#define blastall_language "C"
#define blastall_language2 "C"
#define blastall_source "ftp://ftp.ncbi.nlm.nih.gov/blast/executables/blast+/2.6.0/ncbi-blast-2.6.0+-src.tar.gz"
#define blastall_update_action "never"
#define blastall_mode "psicoffee,expresso,3dcoffee"
#define NCBIBLAST_4_TCOFFEE "legacy_blast.pl"
#define NCBIBLAST_type "protein_homology_predictor"
#define NCBIBLAST_ADDRESS "ftp://ftp.ncbi.nih.gov/blast/executables/LATEST"
#define NCBIBLAST_language "C"
#define NCBIBLAST_language2 "C"
#define NCBIBLAST_source "ftp://ftp.ncbi.nlm.nih.gov/blast/executables/blast+/2.6.0/ncbi-blast-2.6.0+-src.tar.gz"
#define NCBIBLAST_update_action "never"
#define NCBIBLAST_mode "psicoffee,expresso,3dcoffee"
#define SOAPLITE_4_TCOFFEE "SOAP::Lite"
#define SOAPLITE_type "library"
#define SOAPLITE_ADDRESS "http://cpansearch.perl.org/src/MKUTTER/SOAP-Lite-0.710.08/Makefile.PL"
#define SOAPLITE_language "Perl"
#define SOAPLITE_language2 "Perl"
#define SOAPLITE_source "empty"
#define SOAPLITE_update_action "never"
#define SOAPLITE_mode "none"
#define XMLSIMPLE_4_TCOFFEE "XML::Simple"
#define XMLSIMPLE_type "library"
#define XMLSIMPLE_ADDRESS "http://search.cpan.org/~grantm/XML-Simple-2.18/lib/XML/Simple.pm"
#define XMLSIMPLE_language "Perl"
#define XMLSIMPLE_language2 "Perl"
#define XMLSIMPLE_source "empty"
#define XMLSIMPLE_mode "psicoffee,expresso,accurate"
#define x3dna_4_TCOFFEE "x3dna"
#define x3dna_type "RNA_secondarystructure_predictor"
#define x3dna_ADDRESS "http://x3dna.bio.columbia.edu/"
#define x3dna_source "http://www.tcoffee.org/Packages/mirrors/source/x3dna-v2.3-linux-64bit.tar.gz"
#define x3dna_mode "saracoffee"
#define x3dna_update_action "never"
#define FSA_4_TCOFFEE "fsa"
#define FSA_type "sequence_multiple_aligner"
#define FSA_ADDRESS "http://fsa.sourceforge.net/"
#define FSA_language "C++"
#define FSA_language2 "CXX"
#define FSA_source "http://sourceforge.net/projects/fsa/files/fsa-1.15.3.tar.gz/download/"
#define FSA_mode "mcoffee"
#define FSA_version "1.15.3"
#define FSA_update_action "never"
#define MUS4_4_TCOFFEE "mus4"
#define MUS4_type "sequence_multiple_aligner"
#define MUS4_ADDRESS "http://www.drive5.com/muscle/"
#define MUS4_language "C++"
#define MUS4_language2 "GPP"
#define MUS4_source "http://www.drive5.com/muscle/muscle4.0_src.tar.gz"
#define MUS4_mode "mcoffee,rcoffee"
#define MUS4_version "4.0"
#define MUS4_update_action "never"
//TclinkdbEnd
/*New Methods*/
/********************************************/
/*            Various Methoids              */
/********************************************/
#define METHODS_4_TCOFFEE  "~/.t_coffee/methods/"
#define METHOD_4_MSA_WEIGHTS "petra_weight"
/********************************************/
/*            SEQAN LIBRARY                  */
/********************************************/
#define SEQAN_TCOFFEE_4_TCOFFEE "seqan_tcoffee"
/********************************************/
/*            REFORMATING  AND UTILITIES                 */
/********************************************/
#define WGET_4_TCOFFEE "wget"
#define WGET_ADDRESS "http://www.gnu.org/software/wget/"

#define CURL_4_TCOFFEE "curl"
#define CURL_ADDRESS "http://curl.haxx.se/"

#define SEQ_REFORMAT_4_TCOFFEE          "seq_reformat"
#define PS2PDF                          "ps2pdf"
#define EXTRACT_FROM_PDB_4_TCOFFEE      "extract_from_pdb"
#define BLAST_ALN2FASTA_ALN             "blast_aln2fasta_aln.pl"
#define FASTA_ALN2FASTA_ALN_UNIQUE_NAME "fasta_aln2fasta_aln_unique_name.pl"
#define MSF_ALN2FASTA_ALN               "msf_aln2fasta_aln.pl"
#define SEQ2MSA_WEIGHT        "seq2msa_weight"
/********************************************/
/*            DEPRECATED DEF                */
/********************************************/
//Deprecated definitions
#define SIB_BLAST_4_TCOFFEE "blastall.remote"
#define LOCAL_BLAST_4_TCOFFEE "blastall"
#define BLAST_DB_4_TCOFFEE "nr"
#define NCBI_BLAST_4_TCOFFEE ""
/********************************************/
/*            PARAMETER_FILE                */
/********************************************/




/*               PARAMETER FILES               */
#define COLOR_FILE         "seq_reformat.color"
/*This file specifies the 10 colors available to seq_reformat.
If the file is not on the system, hard coded defaults will be used.
The format is as follow:

-------------------------------------------------------------------------------------------
<Your comments (as many lines as needed >
*
<number in the range 0-9> <HTML code> <R value (float)> <G value (float)> <B value (float)>
-------------------------------------------------------------------------------------------
the RGB values are used for the post-script generation, the html code is used in html documents.
*/
